src/main/java/ru/curs/xylophone/MergeRegionContainer.java

```diff
@@ -0,0 +1,230 @@
+
+        if (intersectedRegion.isPresent()) {
+            mergeIntersectedRegions(intersectedRegion.get(), res);
+        } else {
+            mergedRegions.add(new MutableCellRangeAddress(res));
+        }
+    }
+
+    /**
+     * Add merged region.
+     * Static merge or iteration merge.
+     * @param res merge region
+     */
+    public void addMergedRegion(CellRangeAddress res) {
+        Optional<MutableCellRangeAddress> intersectedRegion = findIntersectedRegion(res);
+        if (intersectedRegion.isPresent()) {
+            mergeIntersectedRegions(intersectedRegion.get(), res);
+        } else {
+            mergedRegions.add(new MutableCellRangeAddress(res));
+        }
+    }
+
+    /**
+     * Apply all merged regions to sheet.
+     * @param sheet sheet of docs
+     */
+    public void apply(Sheet sheet) {
+        if (sheet == null) {
+            return;
+        }
+        mergedRegions.stream().map(MutableCellRangeAddress::toCellRangeAddress).forEach(sheet::addMergedRegion);
+        clear();
+    }
+
+    /**
+     * Find main region with current cell.
+     * For defining value in region.
+     * @param res range of current merge region.
+     * @return Current merge region if intersects is false; otherwise - intersected existing region.
+     */
+    public CellRangeAddress findIntersectedRange(CellRangeAddress res) {
+        if (res.getFirstColumn() < 0 || res.getFirstRow() < 0) {
+            throw new IllegalArgumentException("Cannot merge first row with upper cell or first column with left cell");
+        }
+        Optional<MutableCellRangeAddress> intersected = findIntersectedRegion(res);
+        if (intersected.isPresent()) {
+            return intersected.get().toCellRangeAddress();
+        } else {
+            return res;
+        }
+    }
+
+    /**
+     * Clear list.
+     */
+    private void clear() {
+        mergedRegions.clear();
+    }
+
+    /**
+     * Merge current cell with existing region.
+     * @param mergedRegion existing merged region.
+     * @param res current cell with it's neighbour (up or left)
+     */
+    private void mergeIntersectedRegions(MutableCellRangeAddress mergedRegion, CellRangeAddress res) {
+        if (isValidUnion(mergedRegion.toCellRangeAddress(), res)) {
+            CellRangeAddress unionRangeRes = new CellRangeAddress(mergedRegion.getFirstRow(), res.getLastRow(),
+                    mergedRegion.getFirstColumn(), res.getLastColumn());
+            mergedRegion.setAddress(unionRangeRes);
+        }
+    }
+
+    /**
+     * Check if is it valid union.
+     * @param mergedRegion existing merged region
+     * @param result region that will be concatenated.
+     * @return True if it is one row (or column); otherwise - false.
+     */
+    private boolean isValidUnion(CellRangeAddress mergedRegion, CellRangeAddress result) {
+        int startRow = mergedRegion.getFirstRow();
```

Comment:
As far as I understood this method only allows appending one-row-heigh or one-column-wide region.
What about generalization?